### ISSUE TITLE
CUPTI tracer's flat 100k record cap silently truncates activity buffers and drops all device activity

### ISSUE THREAD

**Issue by lvyufeng** [flagos-ai/Torch-FL#412]
## Issue Type
- [x] Bug Report

## AI Agent Information
- **Agent**: Claude Code CLI
- **Model**: Claude Opus 5
- **Session Context**: Filed while fixing #411 (PPU compiled the unavailable device tracer instead of the CUPTI one). With the tracer selection corrected, PPU traces still carried no device activity, and measuring the buffer drain found a second, independent defect in the shared CUPTI tracer.

## Summary

`CuptiDeviceTracer::processBuffer()` (`csrc/profiler/cupti_device_tracer.cc`) bounds its
record loop with a flat record count, `kMaxRecordsPerBuffer = 100000`. That number is not a
property of the buffer: the number of records a CUPTI activity buffer can legitimately hold
depends on the vendor's record mix and the buffer size. PPU emits an external-correlation
record plus a driver record per runtime call, so a single 3.4 MB buffer from a trivial
workload holds 107,829 records -- more than the cap. The loop therefore exits mid-buffer, and
because the device records sit behind that flood of correlation/driver records, **every
kernel, memcpy and memset activity in the buffer is dropped from the trace**. `processBuffer()`
reported it as a possible `cuptiActivityGetNextRecord` infinite loop, which it is not: the
same buffer drains to its last byte once the bound is raised.

## Environment (for bug reports)
<details>
<summary>Click to expand environment details</summary>

- **Platform**: PPU
- **Python**: 3.12.13
- **PyTorch**: 2.10.0 (`torch.version.cuda == "13.0"`)
- **torch_fl**: 0.1.0+ppu, built with `FLAGOS_ACCELERATOR=ppu` from this branch
- **Hardware**: 16x PPU-ZW810E; `PPU_SDK=/usr/local/PPU_SDK`, `CUDA_HOME=$PPU_SDK/CUDA_SDK`

**Build config:**
```bash
export PPU_SDK=/usr/local/PPU_SDK FLAGOS_ACCELERATOR=ppu
export CUDA_HOME=$PPU_SDK/CUDA_SDK
export FLAGOS_BUILD_FLAGGEMS_CPP=0 FLAGOS_SKIP_CUDA_ASSETS=1
pip install --no-build-isolation --no-deps -e . -v
```

**Runtime config:**
```bash
export FLAGOS_TRACE=1            # enables the tracer's own diagnostics
export FLAGOS_DISABLE_CUDA_ASSETS=1
```
</details>

## Reproduction

```python
import torch
import torch_fl  # noqa: F401  (registers the flagos PrivateUse1 backend)

x = torch.randn(512, 512, device="flagos:0")
y = torch.randn(512, 512, device="flagos:0")
with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.PrivateUse1,
    ]
) as prof:
    z = x @ y
    z.sum().item()

print(Counter(e.category for e in prof.events()))
```

## Expected vs Actual Behavior

**Expected:** the trace carries the device work the workload performed -- the two
`at::native` kernels (`gemm`, `sum`), the D2H copy from `z.sum().item()`, and the runtime
calls that issued them.

**Actual:**

```
[flagos] processBuffer exceeded max record count (100000) for buffer size 3451352 bytes;
         possible ActivityGetNextRecord infinite loop
[flagos] session stop: drained 10 device events
[flagos] processTrace events by kind: kernel=0 memcpy=0 memset=0 runtime=10
```

```
categories: ['Trace', 'cpu_op', 'privateuse1_runtime']
```

`kernel=0 memcpy=0 memset=0`: the traced matmul produced no kernel event at all. Only the
first 10 runtime records -- the ones that happened to land before the cap -- survived.

## Root Cause Analysis

Measured with temporary diagnostics in `processBuffer()` that dump the raw buffer, the
record-kind histogram, and how many bytes the iterator actually walked:

```
[dbg] validSize=3452632 first words: 39 3 2 0 26156 0 4 16 ...
[dbg] iter 1 status=0 ptr=...010 kind=39
[dbg] iter 2 status=0 ptr=...028 kind=4
[dbg] iterations=107829 kinds: 1x1 10x5 5x71 4x53910 39x53842
[dbg] walked=3452592 of 3452632
```

* The iterator is **not** looping. It terminates on its own: `walked=3452592` is the last
  record of a 3,452,632-byte buffer, 40 bytes short of the end.
* The buffer holds 107,829 records, 99.9% of them `DRIVER` (kind 4) and
  `EXTERNAL_CORRELATION` (kind 39) -- the infrastructure PPU emits per runtime call. The
  actual activity is 5 `CONCURRENT_KERNEL`, 1 `MEMCPY` and 71 `RUNTIME` records, all of
  which land after the first 100,000 records.
* So the flat cap is a truncation, and the "possible infinite loop" wording was a
  misdiagnosis of a buffer that only needed more iterations than the constant allowed.
* The vendor library is capable of emitting all of it: an independent ctypes probe against
  `/usr/local/PPU_SDK/CUDA_SDK/lib64/libcupti.so.13` (callback registration moved after
  `import torch`, since torch's kineto owns the process-global CUPTI callbacks) observed
  `MEMCPYx1 MEMSETx1 RUNTIMEx247 CONCURRENT_KERNELx9`.

The same flat cap is a latent risk on every CUPTI platform, not just PPU: any vendor whose
per-call record overhead grows, or any workload whose trace window is long enough to fill a
buffer with more than 100,000 records, loses its device activity the same silent way.

## Proposed Solution

Derive the bound from the buffer instead of hardcoding a record count. A CUPTI record is at
least one 32-bit word wide and every record lies inside the buffer, so
`validSize / sizeof(uint32_t)` is the most records a well-formed buffer can yield. That makes
the bound a pure anti-infinite-loop backstop -- the loop already ends on a non-`SUCCESS`
status or a non-advancing record pointer -- and it can never truncate.

Fixed in the same change as #411, since PPU only observes this defect once the tracer is
built at all.

## Verification Plan
- [x] Unit test: n/a -- the defect is in a CUDA/CUPTI-only code path that cannot be
      exercised without vendor hardware or libraries.
- [x] Integration test: `tests/integration/test_profiler_contract.py -m profiler` on PPU.
      The `kernel`, `memcpy` and `runtime` cases fail against the flat cap and pass against
      the size-derived bound. Wired into `.github/configs/ppu.yml` by the #411 change.
- [x] Manual verification: `FLAGOS_TRACE=1` trace of the reproducer, comparing the drained
      event count and the per-kind totals before and after.

## Context & Investigation

1. Read `csrc/profiler/cupti_device_tracer.cc`'s `processBuffer()` and the
   `FLAGOS_METAX_MCPTI` scan path next to it.
2. Built with `FLAGOS_ACCELERATOR=ppu` and confirmed the PTX-free build log compiles
   `cupti_device_tracer.cc` and not `unavailable_device_tracer.cc`.
3. Reproduced the empty device timeline, then added temporary diagnostics to distinguish
   "iterator never terminates" from "iterator terminates late" -- the second is what the
   buffer dump shows.
4. Raised the cap to 2,000,000 to confirm the whole buffer is reachable, then replaced the
   constant with the `validSize`-derived bound and re-measured.
5. Cross-checked with a standalone ctypes probe against PPU's `libcupti.so.13` to prove the
   vendor library emits the records, so the loss was ours and not the vendor's.

## Related Code Locations
- `csrc/profiler/cupti_device_tracer.cc:820` -- the flat bound (now `max_records`)
- `csrc/profiler/cupti_device_tracer.cc:824` -- the record loop it bounds
- `csrc/profiler/cupti_device_tracer.cc:1125` -- the diagnostic that misreported it as an infinite loop

## Checklist - AI Agents MUST Complete All
- [x] I have provided complete environment information
- [x] I have included a minimal, self-contained reproducer
- [x] I have included full error output with traceback
- [x] I have analyzed the root cause (not just symptoms)
- [x] I have proposed a specific solution with implementation approach
- [x] I have identified affected code locations with line numbers
- [x] I have described how to verify the fix
- [x] I have checked for duplicate issues
- [x] All text is in **English** (required per CLAUDE.md)
- [x] Code follows project conventions (checked existing code style)
